Traefik-passthrough doc  file (#48556)
keycloak/keycloak: docs/guides/server/traefik-passthrough.adoc

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/options.adoc" as opts>
<#import "/templates/links.adoc" as links>

> <@tmpl.guide title="Traefik with TLS passthrough" summary="Configure Traefik as a TLS passthrough load balancer for {project_name}." includedOptions="proxy-protocol-enabled metrics-enabled health-enabled" tileVisible="false">

Reviewer: I would argue that `http-management-scheme` is a relevant option to be included :)

=== Static configuration (`traefik.yaml`)

The following `traefik.yaml` configures the entry point for TLS traffic and enables the dashboard.

[source,yaml]
----
entryPoints:
  websecure:
    address: ":8443" # <1>

api:
  insecure: true # <2>
----

<1> Defines the entry point on port `8443` where Traefik accepts incoming TCP/TLS connections.
Clients connect to this port, and Traefik forwards the raw TCP stream to the backend.
<2> Enables the Traefik dashboard on port `8080` without authentication.
This is intended for local development and debugging only and must never be used in production.

=== Dynamic configuration (`keycloak.yaml`)

The following `keycloak.yaml` configures the TCP router, load balancer, and health checks.

[source,yaml]
----
tcp:
  routers:
    keycloak-router:
      rule: "HostSNI(`*`)" # <1>
      service: "keycloak-service"
      priority: 100 # <2>
      entryPoints:
        - "websecure"
      tls:
        passthrough: true # <3>

  serversTransports:
    kc-passthrough:
      proxyProtocol:
        version: 2 # <4>

  services:
    keycloak-service:
      loadBalancer:
        serversTransport: kc-passthrough # <5>

>         # Use JSON syntax to have explicit CR/LF encoding for send/expect.

Reviewer: ```suggestion
        # The inline YAML syntax is required so that \r\n is interpreted literally.
```

        healthCheck: { # <6>
          interval: "5s",
          timeout: "3s",
          port: 9000,
          send: "HEAD /health/ready HTTP/1.0\r\n\r\n",
          expect: "HTTP/1.0 200 OK\r\ncontent-type: application/json; charset=UTF-8\r\ncache-control: no-store\r\n\r\n"
        }
        servers:
          - address: "keycloak1:8443" # <7>
          - address: "keycloak2:8443"
----

> <1> Matches all TLS connections regardless of the SNI hostname.

Reviewer: Add a sentence saying this is not recommended in production or if the proxy handles multiple services (not only Keycloak)

These values affect how quickly Traefik detects that a {project_name} instance is shutting down (see <<graceful-shutdown-considerations>>).

`port: 9000`::
Directs health checks to the management port (`9000`), which is separate from the main application port (`8443`).
The management port is configured to use plain HTTP via `--http-management-scheme=http` (see <<keycloak-configuration>>), which is required because Traefik cannot perform an HTTPS health check on a TCP passthrough service.
This is safe because the management port is only reachable on the internal backend network and is not exposed to clients.
For details on the management port, see the <@links.server id="management-interface"/> {section}.

`send` / `expect`::
Since Traefik operates at the TCP layer, health checks work by sending raw bytes and matching the response.
The JSON syntax is required so that `\r\n` is encoded as a literal carriage-return and line-feed rather than a backslash-n string.
+
NOTE: The `expect` string must match the full HTTP response headers exactly, including the `content-type` charset.
Any change to the {project_name} response format may break this check.

> A simple TCP connect check (omitting `send` and `expect`) is less fragile but would not detect split-brain, database connectivity failures, or an overloaded node — it would only detect a dead process.

Reviewer: Shorter sentence 
```suggestion
A simple TCP connect check (omitting `send` and `expect`) is less fragile but only detects a dead process.
It would not detect split-brain, database connectivity failures, or an overloaded node.
```

`--health-enabled true`::
Enables the health check endpoints on the management port (`9000`).
Without this, the health endpoint that Traefik polls will not be available, and Traefik will mark all {project_name} instances as down.

`--metrics-enabled true`::
Enables the metrics endpoints on the management port.
Enabling metrics also enables an additional status check for the database, so it is recommended to enable metrics.

`--http-management-scheme=http`::
Configures the management port (`9000`) to use plain HTTP instead of HTTPS.
Traefik cannot perform an HTTPS health check on a TCP passthrough service, so the management port must use plain HTTP.
This is safe because the management port is only reachable on the internal backend network and is not exposed to clients.

[[graceful-shutdown-considerations]]
== Graceful shutdown considerations

With TLS passthrough, Traefik cannot signal a connection close at the HTTP level.
The health check timing directly determines how long it takes Traefik to detect that a {project_name} instance is shutting down and stop routing new connections to it.

> With the health check settings from the configuration above (`interval: 5s`, `timeout: 3s`), it may take several seconds for Traefik to mark a {project_name} instance as down.

Reviewer: > it may take several seconds

It should be clear and specify the maximum time it takes. 8 seconds, right? Isn't the shutdown time way too large?  We used 2x the max detection time in HAProxy.

Reviewer 2: The shutdown delay is also responsible to gracefully terminate existing TCP connections. With that, it would be IMHO ok to keep it at 30 seconds.

It would be interesting to know how Traefik handles the failing probe in passthrough: When the probe fails, and when there are existing open TCP connections, do they remain open until Keycloak closes them? I would hope it would do that. If it doesn't there isn't much need to wait for those connections to gracefully close.

Author: I will update the wording to specify the exact maximum detection time of 8 seconds (interval: 5s + timeout: 3s).
I kept --shutdown-delay=30s as the example rather than 16s as Traefik in TLS passthrough cannot send a Connection: close signal at the HTTP level to nudge existing keepalive connections to close. I can also add a NOTE clarifying that 16s is the safe minimum if TCP connection draining is not a concern.

=== Dynamic configuration (`keycloak.yaml`)

The following `keycloak.yaml` configures the TCP router, load balancer, and health checks.

[source,yaml]
----
tcp:
  routers:
    keycloak-router:
      rule: "HostSNI(`*`)" # <1>
      service: "keycloak-service"
      priority: 100 # <2>
      entryPoints:
        - "websecure"
      tls:
        passthrough: true # <3>

  serversTransports:
    kc-passthrough:
      proxyProtocol:
        version: 2 # <4>

  services:
    keycloak-service:
      loadBalancer:
        serversTransport: kc-passthrough # <5>
        # The inline YAML syntax is required so that \r\n is interpreted literally.
        healthCheck: { # <6>
          interval: "5s",
          timeout: "3s",
          port: 9000,
          send: "HEAD /health/ready HTTP/1.0\r\n\r\n",
          expect: "HTTP/1.0 200 OK\r\ncontent-type: application/json; charset=UTF-8\r\ncache-control: no-store\r\n\r\n"
        }
        servers:
          - address: "keycloak1:8443" # <7>
          - address: "keycloak2:8443"
----

<1> Matches all TLS connections regardless of the SNI hostname.
This is suitable when Traefik is dedicated to {project_name} only.

> If Traefik handles multiple backend services, use a specific hostname instead of a wildcard (for example, `HostSNI(`keycloak.example.com`)`) to avoid routing unintended traffic to {project_name}.

Reviewer: The rendering is broken. 

<img width="412" height="40" alt="Image" src="https://github.com/user-attachments/assets/56983f00-eabb-44e0-a671-ea1d0e292711" />

The suggestion below seems to fix it
```suggestion
If Traefik handles multiple backend services, use a specific hostname instead of a wildcard (for example, ``HostSNI(`keycloak.example.com`)``) to avoid routing unintended traffic to {project_name}.
```

<2> Ensures this router takes precedence over any other TCP routers that may be defined.
<3> Instructs Traefik to forward the raw TLS stream without terminating it.
{project_name} handles TLS termination directly.

> <4> Enables link:https://doc.traefik.io/traefik/routing/services/#proxy-protocol[PROXY protocol v2] on all backend connections.

Reviewer: Link points to an incorrect page
```suggestion
<4> Enables link:https://doc.traefik.io/traefik/reference/routing-configuration/tcp/serverstransport/#proxyprotocolversion[PROXY protocol v2] on all backend connections.
```

This allows {project_name} to read the real client IP address from the PROXY protocol header added by Traefik.
This requires {project_name} to be configured with `--proxy-protocol-enabled=true` (see <<keycloak-configuration>>).
<5> Links the load balancer to the named transport above, activating PROXY protocol v2 for all connections to the backend servers.

> <6> Configures a link:https://doc.traefik.io/traefik/routing/services/#health-check_1[TCP-level health check] against {project_name}'s management port (`9000`).

Reviewer: incorrect link

```suggestion
<6> Configures a link:https://doc.traefik.io/traefik/reference/routing-configuration/tcp/service/#health-check[TCP-level health check] against {project_name}'s management port (`9000`).
```

This is safe because the management port is only reachable on the internal backend network and is not exposed to clients.
For details on the management port, see the <@links.server id="management-interface"/> {section}.

`send` / `expect`::
Since Traefik operates at the TCP layer, health checks work by sending raw bytes and matching the response.
The JSON syntax is required so that `\r\n` is encoded as a literal carriage-return and line-feed rather than a backslash-n string.
+
NOTE: The `expect` string must match the full HTTP response headers exactly, including the `content-type` charset.
Any change to the {project_name} response format may break this check.
A simple TCP connect check (omitting `send` and `expect`) is less fragile but only detects a dead process.
It would not detect split-brain, database connectivity failures, or an overloaded node.
Native HTTP health checks for TCP services are not yet supported in Traefik; track upstream progress at link:https://github.com/traefik/traefik/pull/12606[traefik/traefik#12606].

> <7> Defines the two {project_name} backend servers.

Reviewer: This is in the wrong place 😅

Traffic is distributed across both instances using round-robin load balancing.

[[keycloak-configuration]]
== {project_name} configuration

With TLS passthrough, {project_name} requires the following configuration:

<@kc.start parameters="--proxy-protocol-enabled true --health-enabled true --metrics-enabled true --http-management-scheme=http"/>

`--proxy-protocol-enabled true`::
Enables the PROXY protocol so that {project_name} can read the real client IP address from the PROXY protocol header added by Traefik.
Do not set `--proxy-headers` when using TLS passthrough because Traefik cannot inject HTTP headers into the encrypted traffic.
The `--proxy-protocol-enabled` option and the `--proxy-headers` option are mutually exclusive.

`--health-enabled true`::
Enables the health check endpoints on the management port (`9000`).
Without this, the health endpoint that Traefik polls will not be available, and Traefik will mark all {project_name} instances as down.

`--metrics-enabled true`::
Enables the metrics endpoints on the management port.
Enabling metrics also enables an additional status check for the database, so it is recommended to enable metrics.

> `--http-management-scheme=http`::
> Configures the management port (`9000`) to use plain HTTP instead of HTTPS.
> Traefik cannot perform an HTTPS health check on a TCP passthrough service, so the management port must use plain HTTP.
> This is safe because the management port is only reachable on the internal backend network and is not exposed to clients.

Reviewer: Let me know if this sounds clearer @ruchikajha95 @ahus1 
```suggestion
Configures the management port (`9000`) to use plain HTTP instead of HTTPS.
Traefik does not support native HTTP health checks for TCP services; it can only send and match raw bytes over a plain TCP connection.
Disabling TLS on the management port allows the raw `send`/`expect` health check (see the health check configuration above) to communicate with {project_name} in plain HTTP.
This is safe because the management port is only reachable on the internal backend network and is not exposed to clients.
```
